### PR TITLE
Skios 145 mutators and resolvers for bookmarks

### PR DESCRIPTION
**Pull request by LukasLJL**
# Skiosa - PR
- [x] added test(s), or explain why one is not needed/possible...
- [x] added comments to all the functions
- [x] added documentation to latex project
- [x] code can't be split up into smaller logical units
- [x] code is formated and linted

### CODE REVIEW

File: src/service/bookmarkService.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import { UserInfo } from "../model/jwt";
+import { PaginationArg } from "../model/paginationArg";
+
+export interface BookmarkService {
+    /**
+     * @author Lukas Huida
+     * @summary add article to bookmarks
+     * @description adds an article to the current user's bookmarks
+     */
+    addBookmark(currentUserInfo: UserInfo, articleId: number): Promise<boolean>;
```
Comment (jeppard) [inline src/service/bookmarkService.ts:11]: Does not match contribution Guidelines

File: src/service/bookmarkService.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    /**
+     * @author Lukas Huida
+     * @summary delete article from bookmarks
+     * @description removes an article from the current user's bookmarks
+     */
+    deleteBookmark(currentUserInfo: UserInfo, articleId: number): Promise<boolean>;
```
Comment (jeppard) [inline src/service/bookmarkService.ts:18]: Also doesn't match contribution guidelines

File: src/service/impl/bookmarkServiceImpl.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                    currentUser.bookmarks?.push(article);
+                    userRepository.manager.save(currentUser);
+                    resolve(true);
+                } else {
+                    reject(new Error("Article does not exist"));
+                }
+            }).catch(() => {
+                reject(new Error("Article does not exist"));
+            });
+        });
```
Comment (jeppard) [inline src/service/impl/bookmarkServiceImpl.ts:35]: unnecessary Promise

File: src/service/impl/bookmarkServiceImpl.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+                    userRepository.manager.save(currentUser);
+                    resolve(true);
+                } else {
+                    reject(new Error("Article does not exist"));
+                }
+            }).catch(() => {
+                reject(new Error("Article does not exist"));
+            });
+        });
```
Comment (jeppard) [inline src/service/impl/bookmarkServiceImpl.ts:63]: Unnecessary Promise

File: src/service/impl/bookmarkServiceImpl.ts [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    async bookmarks(@CurrentUser() currentUserInfo: UserInfo, @Arg("PaginationArg", { nullable: true }) paginated?: PaginationArg): Promise<Article[]> {
+        const userRepository = dataSource.getRepository(User);
+        const currentUser = await getCurrentUser(userRepository, currentUserInfo);
+
+        if (!currentUser.bookmarks || currentUser.bookmarks.length === 0) {
+            return [];
+        }
+
+        return paginate(currentUser.bookmarks, paginated);
+    }
```
Comment (jeppard) [inline src/service/impl/bookmarkServiceImpl.ts:77]: getCurrentUser can throw errors. Awaiting a Promise which can reject is unclean please use .then()

File: src/service/mockup/articleServiceMock.ts [hunk 6]
```diff
@@ -90,4 +92,10 @@ export class ArticleServiceMock extends MockService implements ArticleService {
       throw new Error(`Article with id: ${id} not found!`);
     }
   }
+
+  bookmarkStatus(currentUserInfo: UserInfo, article: Article): Promise<boolean> {
+    console.log(currentUserInfo);
+    console.log(article);
```
Comment (jeppard) [inline src/service/mockup/articleServiceMock.ts:98]: Debug output still present

**Review by jeppard — CHANGES_REQUESTED** [review]

File: src/service/mockup/articleServiceMock.ts (hunk 6 above)
Comment (LukasLJL) [reply src/service/mockup/articleServiceMock.ts:98]: mockservice ->  will be delete in a future pr

Comment (jeppard) [reply src/service/mockup/articleServiceMock.ts:98]: Ok don't know the plans so i guess it's fine

File: src/service/bookmarkService.ts (hunk 1 above)
Comment (LukasLJL) [reply src/service/bookmarkService.ts]: fixed

File: src/service/bookmarkService.ts (hunk 2 above)
Comment (LukasLJL) [reply src/service/bookmarkService.ts]: fixed

File: src/service/impl/bookmarkServiceImpl.ts (hunk 3 above)
Comment (LukasLJL) [reply src/service/impl/bookmarkServiceImpl.ts]: should be fixed

File: src/service/impl/bookmarkServiceImpl.ts (hunk 4 above)
Comment (LukasLJL) [reply src/service/impl/bookmarkServiceImpl.ts]: should be fixed

File: src/service/impl/bookmarkServiceImpl.ts (hunk 5 above)
Comment (LukasLJL) [reply src/service/impl/bookmarkServiceImpl.ts]: should be fixed

**Review by eintim — APPROVED** [review]

File: src/service/impl/bookmarkServiceImpl.ts [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+  @Mutation(() => Boolean)
+  @Authorized("realm:skiosa-user")
+  async deleteBookmark(
+    @CurrentUser() currentUserInfo: UserInfo,
+    @Arg("articleId") articleId: number
+  ): Promise<boolean> {
+    const userRepository = dataSource.getRepository(User);
+
+    const currentUser = await getCurrentUser(userRepository, currentUserInfo);
```
Comment (jeppard) [inline src/service/impl/bookmarkServiceImpl.ts:46]: getCurrentUser is able to throw errors. await is problematic. Errors will be catched if using then

File: src/service/impl/bookmarkServiceImpl.ts [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+          } else {
+            resolve([]);
+          }
+        })
+        .catch(() => {
+          reject("User does not exist");
+        });
+    });
+  }
+}
```
Comment (jeppard) [inline src/service/impl/bookmarkServiceImpl.ts:93]: Dont build new Prommise get CurrentUser is a Promise

**Review by jeppard — CHANGES_REQUESTED** [review]

File: src/service/impl/bookmarkServiceImpl.ts (hunk 8 above)
Comment (jeppard) [reply src/service/impl/bookmarkServiceImpl.ts:93]: or better phrased getCurrentUser returns a promisse

**Review by jeppard — APPROVED** [review]
Thanks looks good!

**Pull request merged**